### PR TITLE
Lock Abstraction WIP

### PR DESCRIPTION
**Pull request by spoon16**
These are the changes that have not been approved for merging into the main Lock Abstraction PR #442

### CODE REVIEW

File: store/src/java/com/zimbra/cs/mailbox/LocalMailboxLock.java [hunk 1]
```diff
@@ -43,233 +35,47 @@
  *
  */
 public final class LocalMailboxLock implements MailboxLock {
```
Comment (spoon16) [inline store/src/java/com/zimbra/cs/mailbox/LocalMailboxLock.java]: You deleted a ton of code from this class. I've not read through it all, but I'm worried that you changed the semantics of how this class works.

File: store/src/java/com/zimbra/cs/service/mail/GetCalendarItem.java [hunk 2]
```diff
@@ -90,8 +91,8 @@ public Element handle(Element request, Map<String, Object> context) throws Servi
         }
         Element response = getResponseElement(zsc);
         CalendarItem calItem = null;
-        mbox.lock.lock(false);
-        try {
+        try (final MailboxLock l = mbox.lockFactory.readLock()) {
```
Comment (spoon16) [inline store/src/java/com/zimbra/cs/service/mail/GetCalendarItem.java]: Change to `mbox.lock(false)` instead of `mbox.lockFactory.readLock()`.

File: store/src/java/com/zimbra/cs/service/mail/ModifyCalendarItem.java [hunk 3]
```diff
@@ -108,8 +109,8 @@ public Element handle(Element request, Map<String, Object> context) throws Servi
         MailSendQueue sendQueue = new MailSendQueue();
         Element response = getResponseElement(zsc);
         int compNum = (int) request.getAttributeLong(MailConstants.A_CAL_COMP, 0);
-        mbox.lock.lock();
-        try {
+        try (final MailboxLock l = mbox.lockFactory.writeLock()) {
```
Comment (spoon16) [inline store/src/java/com/zimbra/cs/service/mail/ModifyCalendarItem.java]: Change this to `mbox.lock(true)` instead of `mbox.lockFactory.writeLock()`

Comment (spoon16) [reply store/src/java/com/zimbra/cs/service/mail/ModifyCalendarItem.java]: Including the mailbox lock in `try-with-resources` here changes the semantics of this operation. Be careful not to do this.

Notice that `sendQueue.send()` will now occur before the `MailboxLock` is closed.

You are going to need to wrap the body of this try statement with another `try-with-resources` that manages the lifecycle of the `MailboxLock`

File: store/src/java/com/zimbra/cs/service/mail/SendMsg.java [hunk 4]
```diff
@@ -660,8 +661,8 @@ private boolean fixupReply(ZVCalendar ical) throws ServiceException {
                     continue;
 
                 String uid = replyInv.getUid();
-                mMailbox.lock.lock();
-                try {
+                try (final MailboxLock l = mMailbox.lockFactory.writeLock()) {
```
Comment (spoon16) [inline store/src/java/com/zimbra/cs/service/mail/SendMsg.java]: Do not use `lockFactory`

File: store/src/java/com/zimbra/cs/service/mail/SetCalendarItem.java [hunk 5]
```diff
@@ -105,8 +106,8 @@ public Element handle(Element request, Map<String, Object> context) throws Servi
         int flags = Flag.toBitmask(request.getAttribute(MailConstants.A_FLAGS, null));
         String[] tags = TagUtil.parseTags(request, mbox, octxt);
 
-        mbox.lock.lock();
-        try {
+        try (final MailboxLock l = mbox.lockFactory.writeLock()) {
```
Comment (spoon16) [inline store/src/java/com/zimbra/cs/service/mail/SetCalendarItem.java]: Do not use `lockFactory`

File: store/src/java/com/zimbra/cs/mailbox/LocalMailboxLock.java (hunk 1 above)
Comment (spoon16) [reply store/src/java/com/zimbra/cs/mailbox/LocalMailboxLock.java]: There is a lot of logic in Zimbra's lock/release methods that looks important, like we don't want to change that. It's critical that this class behave EXACTLY as it did before since this is the default implementation of locking in the Zimbra product and needs to be 100% backwards compatible.

File: store/src/java/com/zimbra/cs/datasource/imap/FolderChanges.java [hunk 6]
```diff
@@ -47,16 +48,14 @@ private FolderChanges findChanges(int lastSync) throws ServiceException {
         List<Integer> tombstones;
         List<Folder> modifiedFolders;
 
-        mbox.lock.lock(false);
-        try {
+        try (final MailboxLock l = mbox.lockFactory.writeLock()) {
```
Comment (spoon16) [inline store/src/java/com/zimbra/cs/datasource/imap/FolderChanges.java]: do not use lockFactory

File: store/src/java/com/zimbra/cs/account/ShareInfo.java [hunk 7]
```diff
@@ -311,17 +312,15 @@ private String getKey(String ownerAcctId, int remoteFolderId) {
 
             Map<String, Integer> mountpoints = new HashMap<String, Integer>();
 
-            mbox.lock.lock(false);
-            try {
+            try (final MailboxLock l = mbox.lockFactory.readLock()) {
```
Comment (spoon16) [inline store/src/java/com/zimbra/cs/account/ShareInfo.java]: do not use lockFactory

File: store/src/java/com/zimbra/cs/datasource/imap/ImapFolderSync.java [hunk 8]
```diff
@@ -388,13 +389,11 @@ private boolean isSyncEnabled() throws ServiceException {
 
     private FolderSyncState newSyncState() throws ServiceException {
         FolderSyncState ss = new FolderSyncState();
-        mailbox.lock.lock();
-        try {
+        try (final MailboxLock l = mailbox.lockFactory.writeLock()) {
```
Comment (spoon16) [inline store/src/java/com/zimbra/cs/datasource/imap/ImapFolderSync.java]: do not use lockFactory

File: store/src/java/com/zimbra/cs/datasource/imap/MessageChanges.java [hunk 9]
```diff
@@ -52,16 +53,14 @@ private MessageChanges findChanges(int changeId) throws ServiceException {
         List<Integer> tombstones;
         List<Integer> modifiedItems;
 
-        mbox.lock.lock(false);
-        try {
+        try (final MailboxLock l = mbox.lockFactory.readLock()) {
```
Comment (spoon16) [inline store/src/java/com/zimbra/cs/datasource/imap/MessageChanges.java]: do not use lockFactory

File: store/src/java/com/zimbra/cs/dav/resource/MailItemResource.java [hunk 10]
```diff
@@ -512,19 +513,18 @@ public void patchProperties(DavContext ctxt, java.util.Collection<Element> set,
         Mailbox mbox = null;
         try {
             mbox = getMailbox(ctxt);
-            mbox.lock.lock();
-            Metadata data = mbox.getConfig(ctxt.getOperationContext(), CONFIG_KEY);
-            if (data == null) {
-                data = new Metadata();
+            try (final MailboxLock l = mbox.lock(true)) {
```
Comment (spoon16) [inline store/src/java/com/zimbra/cs/dav/resource/MailItemResource.java]: You have changed the lifecycle of the lock. Needs to be fixed.

You can see from the old code that the lock is not released until after the `finally` block of the parent `try/catch`

File: store/src/java/com/zimbra/cs/db/DbDataSource.java [hunk 11]
```diff
@@ -339,8 +340,8 @@ public static void deleteMapping(DataSource ds, int itemId, boolean isBatch) thr
 
         ZimbraLog.datasource.debug("Deleting all mappings for dataSource %s in folder %d", ds.getName(), folderId);
 
-        mbox.lock.lock();
-        try {
+        try (final MailboxLock l = mbox.lockFactory.writeLock()) {
```
Comment (spoon16) [inline store/src/java/com/zimbra/cs/db/DbDataSource.java]: do not use `lockFactory.

File: store/src/java/com/zimbra/cs/imap/LocalImapMailboxStore.java [hunk 12]
```diff
@@ -137,17 +134,15 @@ public InputStreamWithSize getByImapId(OperationContext octxt, int imapId, Strin
 
     @Override
     public void checkAppendMessageFlags(OperationContext octxt, List<AppendMessage> appends) throws ServiceException {
-        mailbox.lock.lock();
-        try {
+        try (final MailboxLock l = mailbox.lockFactory.writeLock()) {
```
Comment (spoon16) [inline store/src/java/com/zimbra/cs/imap/LocalImapMailboxStore.java]: do not use `lockFactory`

File: store/src/java/com/zimbra/cs/mailbox/ContactBackupThread.java [hunk 13]
```diff
@@ -298,10 +299,10 @@ private void purgeOldBackups(OperationContext octxt, Mailbox mbox, Folder folder
         if (!list.isEmpty()) {
             int counter = 0;
             for (Integer id : list.getAllIds()) {
-                try {
-                    mbox.beginTransaction(OPERATION, octxt);
+                try (final MailboxLock l = mbox.lockFactory.writeLock();
```
Comment (spoon16) [inline store/src/java/com/zimbra/cs/mailbox/ContactBackupThread.java]: do not use `lockFactory`

File: store/src/java/com/zimbra/cs/mailbox/MailboxIndex.java [hunk 14]
```diff
@@ -439,8 +436,8 @@ void indexAllDeferredFlagItems() throws ServiceException {
      * Mailbox version (1.0,1.1)->1.2 Re-Index all contacts.
      */
     void upgradeMailboxTo1_2() throws ServiceException {
-        mailbox.lock.lock();
-        try {
+        try (final MailboxLock l = mailbox.lockFactory.writeLock()) {
```
Comment (spoon16) [inline store/src/java/com/zimbra/cs/mailbox/MailboxIndex.java]: do not use `lockFactory`

File: store/src/java/com/zimbra/cs/mailbox/MailboxIndex.java [hunk 15]
```diff
@@ -449,22 +446,18 @@ void upgradeMailboxTo1_2() throws ServiceException {
...
         }
     }
 
     /**
      * Entry point for Redo-logging system only. Everybody else should use queueItemForIndexing inside a transaction.
      */
     public void redoIndexItem(MailItem item) {
-        mailbox.lock.lock();
-        try {
+        try (final MailboxLock l = mailbox.lockFactory.writeLock()) {
```
Comment (spoon16) [inline store/src/java/com/zimbra/cs/mailbox/MailboxIndex.java]: do not use `lockFactory`

File: store/src/java/com/zimbra/cs/mailbox/MailboxIndex.java [hunk 16]
```diff
@@ -655,9 +648,8 @@ public boolean isCompactIndexInProgress() {
     public List<DbSearch.Result> search(DbSearchConstraints constraints,
             DbSearch.FetchMode fetch, SortBy sort, int offset, int size, boolean inDumpster) throws ServiceException {
         List<DbSearch.Result> result;
-        boolean success = false;
-        try {
-            mailbox.beginReadTransaction("search", null);
+            try (final MailboxLock l = mailbox.lockFactory.writeLock();
```
Comment (spoon16) [inline store/src/java/com/zimbra/cs/mailbox/MailboxIndex.java]: do not use `lockFactory`

File: store/src/java/com/zimbra/cs/mailbox/MailboxManager.java [hunk 17]
```diff
@@ -888,19 +890,28 @@ private synchronized Mailbox createMailboxInternal(OperationContext octxt, Accou
                     instantiateExternalVirtualMailbox(data) : instantiateMailbox(data);
             mbox.setGalSyncMailbox(isGalSyncAccount);
             // the existing Connection is used for the rest of this transaction...
-            mbox.beginTransaction("createMailbox", octxt, redoRecorder, conn);
-
-            if (created) {
-                // create the default folders
-                mbox.initialize();
-            }
+            try (final MailboxLock l = mbox.lockFactory.writeLock();
```
Comment (spoon16) [inline store/src/java/com/zimbra/cs/mailbox/MailboxManager.java]: do not use `lockFactory`

File: store/src/java/com/zimbra/cs/mailbox/Message.java [hunk 18]
```diff
@@ -1446,8 +1447,8 @@ boolean move(Folder target) throws ServiceException {
 
     public ParsedMessage getParsedMessage() throws ServiceException {
         ParsedMessage pm = null;
-        mMailbox.lock.lock();
-        try {
+        try (final MailboxLock l = mMailbox.lockFactory.writeLock()) {
```
Comment (spoon16) [inline store/src/java/com/zimbra/cs/mailbox/Message.java]: do not use `lockFactory`

File: store/src/java/com/zimbra/cs/mailbox/ContactBackupThread.java [hunk 19]
```diff
@@ -298,10 +299,10 @@ private void purgeOldBackups(OperationContext octxt, Mailbox mbox, Folder folder
...
+                     final Mailbox.MailboxTransaction t = new Mailbox(mbox.getData()).new MailboxTransaction(OPERATION, octxt, l)) {
```
Comment (spoon16) [inline store/src/java/com/zimbra/cs/mailbox/ContactBackupThread.java]: Replace with.

```java
final Mailbox.MailboxTransaction t = mbox.new MailboxTransaction(OPERATION, octxt, l)
```

File: store/src/java/com/zimbra/cs/mailbox/MailboxIndex.java [hunk 20]
```diff
@@ -393,11 +394,9 @@ public boolean verify(PrintStream out) throws ServiceException {
     @SuppressWarnings("deprecation")
     void indexAllDeferredFlagItems() throws ServiceException {
         try {
-            mailbox.lock.lock();
-            try {
-                boolean success = false;
+            try (final MailboxLock l = mailbox.lockFactory.writeLock()) {
                 try {
-                    mailbox.beginTransaction("indexAllDeferredFlagItems", null);
+                    final Mailbox.MailboxTransaction t = new Mailbox(mailbox.getData()).new MailboxTransaction("indexAllDeferredFlagItems", null, l);
```
Comment (spoon16) [inline store/src/java/com/zimbra/cs/mailbox/MailboxIndex.java]: Replace with.

```java
final Mailbox.MailboxTransaction t = mbox.new MailboxTransaction(OPERATION, octxt, l)
```

File: store/src/java/com/zimbra/cs/mailbox/MailboxIndex.java [hunk 21]
```diff
@@ -655,9 +648,8 @@ public boolean isCompactIndexInProgress() {
     public List<DbSearch.Result> search(DbSearchConstraints constraints,
             DbSearch.FetchMode fetch, SortBy sort, int offset, int size, boolean inDumpster) throws ServiceException {
         List<DbSearch.Result> result;
-        boolean success = false;
-        try {
-            mailbox.beginReadTransaction("search", null);
+            try (final MailboxLock l = mailbox.lockFactory.writeLock();
+                 final Mailbox.MailboxTransaction t = new Mailbox(mailbox.getData()).new MailboxTransaction("search", null, l)) {
```
Comment (spoon16) [inline store/src/java/com/zimbra/cs/mailbox/MailboxIndex.java]: Replace with.

```java
final Mailbox.MailboxTransaction t = mbox.new MailboxTransaction(OPERATION, octxt, l)
```

File: store/src/java/com/zimbra/cs/mailbox/MailboxManager.java [hunk 22]
```diff
@@ -888,19 +890,28 @@ private synchronized Mailbox createMailboxInternal(OperationContext octxt, Accou
...
+                 final Mailbox.MailboxTransaction t = new Mailbox(mbox.getData()).new MailboxTransaction("createMailbox", octxt,l,redoRecorder, conn)) {
```
Comment (spoon16) [inline store/src/java/com/zimbra/cs/mailbox/MailboxManager.java]: Replace with.

```java
final Mailbox.MailboxTransaction t = mbox.new MailboxTransaction(OPERATION, octxt, l)
```

File: store/src/java/com/zimbra/cs/mailbox/MailboxManager.java [hunk 23]
```diff
@@ -888,19 +890,28 @@ private synchronized Mailbox createMailboxInternal(OperationContext octxt, Accou
...
+                if (created) {
```
Comment (spoon16) [inline store/src/java/com/zimbra/cs/mailbox/MailboxManager.java]: remove formatting. only commit actual changes

File: store/src/java/com/zimbra/cs/mailbox/MailboxIndex.java (hunk 20 above)
Comment (spoon16) [reply store/src/java/com/zimbra/cs/mailbox/MailboxIndex.java]: You may also consider adding back a method to mailbox like this.

```java
public MailboxTransaction beginTransaction(args) {
  return new MailboxTransaction(args);
}
```

File: store/src/java/com/zimbra/cs/mailbox/LocalMailboxLock.java (hunk 1 above)
Comment (Raffaell0) [reply store/src/java/com/zimbra/cs/mailbox/LocalMailboxLock.java]: redirected old lock implementation, please take a look

File: store/src/java/com/zimbra/cs/dav/resource/MailItemResource.java [hunk 24]
```diff
@@ -434,7 +434,7 @@ public void patchProperties(DavContext ctxt, java.util.Collection<Element> set,
                     mbox.rename(ctxt.getOperationContext(), mId, type, val, mFolderId);
                     setProperty(DavElements.P_DISPLAYNAME, val);
                     UrlNamespace.addToRenamedResource(getOwner(), uri, this);
-                    UrlNamespace.addToRenamedResource(getOwner(), uri.substring(0, uri.length()-1), this);
+                    UrlNamespace.addToRenamedResource(getOwner(), uri.substring(0, uri.length() - 1), this);
```
Comment (spoon16) [inline store/src/java/com/zimbra/cs/dav/resource/MailItemResource.java]: please do not commit formatting changes in this PR

File: store/src/java/com/zimbra/cs/dav/resource/MailItemResource.java [hunk 25]
```diff
@@ -451,7 +451,7 @@ public void patchProperties(DavContext ctxt, java.util.Collection<Element> set,
                     color.setColor(col);
                 try {
                     Mailbox mbox = getMailbox(ctxt);
-                    mbox.setColor(ctxt.getOperationContext(), new int[] { mId }, type, color);
+                    mbox.setColor(ctxt.getOperationContext(), new int[]{mId}, type, color);
```
Comment (spoon16) [inline store/src/java/com/zimbra/cs/dav/resource/MailItemResource.java]: formatting change

File: store/src/java/com/zimbra/cs/dav/resource/MailItemResource.java [hunk 26]
```diff
@@ -480,7 +480,7 @@ public void patchProperties(DavContext ctxt, java.util.Collection<Element> set,
                         // Update the view for this collection. This collection may get cached if display name is modified.
                         // See UrlNamespace.addToRenamedResource()
                         if (this instanceof Collection) {
-                            ((Collection)this).view = type;
+                            ((Collection) this).view = type;
```
Comment (spoon16) [inline store/src/java/com/zimbra/cs/dav/resource/MailItemResource.java]: formatting change

File: store/src/java/com/zimbra/cs/dav/resource/MailItemResource.java [hunk 27]
```diff
@@ -497,7 +497,7 @@ public void patchProperties(DavContext ctxt, java.util.Collection<Element> set,
         if (mDeadProps.size() > 0) {
             org.dom4j.Document doc = org.dom4j.DocumentHelper.createDocument();
             Element top = doc.addElement(CONFIG_KEY);
-            for (Map.Entry<QName,Element> entry : mDeadProps.entrySet())
+            for (Map.Entry<QName, Element> entry : mDeadProps.entrySet())
```
Comment (spoon16) [inline store/src/java/com/zimbra/cs/dav/resource/MailItemResource.java]: formatting change

File: store/src/java/com/zimbra/cs/dav/resource/MailItemResource.java [hunk 28]
```diff
@@ -507,25 +507,34 @@ public void patchProperties(DavContext ctxt, java.util.Collection<Element> set,
             configVal = new String(out.toByteArray(), "UTF-8");
 
             if (configVal.length() > PROP_LENGTH_LIMIT)
-                for (Map.Entry<QName,Element> entry : mDeadProps.entrySet())
+                for (Map.Entry<QName, Element> entry : mDeadProps.entrySet())
```
Comment (spoon16) [inline store/src/java/com/zimbra/cs/dav/resource/MailItemResource.java]: formatting change

File: store/src/java/com/zimbra/cs/dav/resource/MailItemResource.java [hunk 29]
```diff
@@ -507,25 +507,34 @@ public void patchProperties(DavContext ctxt, java.util.Collection<Element> set,
...
                     ctxt.getResponseProp().addPropError(entry.getKey(), new DavException("prop length exceeded", DavProtocol.STATUS_INSUFFICIENT_STORAGE));
         }
         Mailbox mbox = null;
+        MailboxLock l = null;
         try {
             mbox = getMailbox(ctxt);
-            try (final MailboxLock l = mbox.lock(true)) {
+            try  {
+                l = mbox.lock(true);
```
Comment (spoon16) [inline store/src/java/com/zimbra/cs/dav/resource/MailItemResource.java]: You need to check if `mbox == null` before calling `mbox.lock`.

File: store/src/java/com/zimbra/cs/dav/resource/MailItemResource.java [hunk 30]
```diff
@@ -507,25 +507,34 @@ public void patchProperties(DavContext ctxt, java.util.Collection<Element> set,
...
                 data.put(Integer.toString(mId), configVal);
                 mbox.setConfig(ctxt.getOperationContext(), CONFIG_KEY, data);
+            }finally {
+                
             }
         } catch (ServiceException se) {
             for (QName qname : reqProps)
                 ctxt.getResponseProp().addPropError(qname, new DavException(se.getMessage(), HttpServletResponse.SC_FORBIDDEN));
+        } finally {
+            if (mbox != null)
```
Comment (spoon16) [inline store/src/java/com/zimbra/cs/dav/resource/MailItemResource.java]: Here you can check if `l == null`

File: store/src/java/com/zimbra/cs/service/mail/CreateCalendarItemException.java [hunk 31]
```diff
@@ -107,8 +108,8 @@ public Element handle(Element request, Map<String, Object> context) throws Servi
 
         MailSendQueue sendQueue = new MailSendQueue();
         Element response = getResponseElement(zsc);
-        mbox.lock.lock();
-        try {
+        try (final MailboxLock l = mbox.lock(true)) {
```
Comment (spoon16) [inline store/src/java/com/zimbra/cs/service/mail/CreateCalendarItemException.java]: This causes `sendQueue.send();` to be executed within the lock. You need to make sure that it's executed outside of the lock which means you can't use `try-with-resources` here.

File: store/src/java/com/zimbra/cs/dav/resource/MailItemResource.java (hunk 24 above)
Comment (spoon16) [reply store/src/java/com/zimbra/cs/dav/resource/MailItemResource.java]: The formatting change is still here?

File: store/src/java/com/zimbra/cs/dav/resource/MailItemResource.java (hunk 25 above)
Comment (spoon16) [reply store/src/java/com/zimbra/cs/dav/resource/MailItemResource.java]: the formatting change is still here

File: store/src/java/com/zimbra/cs/mailbox/MailboxIndex.java [hunk 32]
```diff
@@ -393,11 +394,9 @@ public boolean verify(PrintStream out) throws ServiceException {
     @SuppressWarnings("deprecation")
     void indexAllDeferredFlagItems() throws ServiceException {
         try {
-            mailbox.lock.lock();
-            try {
-                boolean success = false;
+            try (final MailboxLock l = mailbox.lock(true)) {
                 try {
-                    mailbox.beginTransaction("indexAllDeferredFlagItems", null);
+                    final Mailbox.MailboxTransaction t = mailbox.new MailboxTransaction("indexAllDeferredFlagItems", null, l);
```
Comment (spoon16) [inline store/src/java/com/zimbra/cs/mailbox/MailboxIndex.java]: this should be in `try-with-resources`

File: store/src/java/com/zimbra/cs/mailbox/MailboxIndex.java [hunk 33]
```diff
@@ -676,9 +668,9 @@ public boolean isCompactIndexInProgress() {
                     }
                 }
             }
-            success = true;
+            t.commit();
         } finally {
-            mailbox.endTransaction(success);
+
```
Comment (spoon16) [inline store/src/java/com/zimbra/cs/mailbox/MailboxIndex.java]: Remove empty `finally`

File: store/src/java/com/zimbra/cs/mailbox/MailboxIndex.java [hunk 34]
```diff
@@ -414,9 +413,9 @@ void indexAllDeferredFlagItems() throws ServiceException {
                     }
                     mailbox.getOperationConnection(); // we must call this before DbMailItem.alterTag
                     DbTag.alterTag(indexingDeferredFlag, deferredTagsToClear, false);
-                    success = true;
+                    t.commit();
                 } finally {
-                    mailbox.endTransaction(success);
+
```
Comment (spoon16) [inline store/src/java/com/zimbra/cs/mailbox/MailboxIndex.java]: Remove empty `finally`

File: store/src/java/com/zimbra/cs/dav/resource/MailItemResource.java [hunk 35]
```diff
@@ -506,26 +507,34 @@ public void patchProperties(DavContext ctxt, java.util.Collection<Element> set,
...
+            }finally {
+                
             }
-            data.put(Integer.toString(mId), configVal);
-            mbox.setConfig(ctxt.getOperationContext(), CONFIG_KEY, data);
         } catch (ServiceException se) {
             for (QName qname : reqProps)
                 ctxt.getResponseProp().addPropError(qname, new DavException(se.getMessage(), HttpServletResponse.SC_FORBIDDEN));
         } finally {
             if (mbox != null)
```
Comment (spoon16) [inline store/src/java/com/zimbra/cs/dav/resource/MailItemResource.java]: check `l` instead of `mbox`

File: store/src/java/com/zimbra/cs/dav/resource/MailItemResource.java [hunk 36]
```diff
@@ -506,26 +507,34 @@ public void patchProperties(DavContext ctxt, java.util.Collection<Element> set,
...
         Mailbox mbox = null;
+        MailboxLock l = null;
         try {
             mbox = getMailbox(ctxt);
-            mbox.lock.lock();
-            Metadata data = mbox.getConfig(ctxt.getOperationContext(), CONFIG_KEY);
-            if (data == null) {
-                data = new Metadata();
+            try  {
+                l = mbox.lock(true);
```
Comment (spoon16) [inline store/src/java/com/zimbra/cs/dav/resource/MailItemResource.java]: before locking you need to check if `mbox == null`

File: store/src/java/com/zimbra/cs/mailbox/MailboxManager.java [hunk 37]
```diff
@@ -888,19 +890,28 @@ private synchronized Mailbox createMailboxInternal(OperationContext octxt, Accou
...
-            cacheMailbox(mbox);
-            redoRecorder.setMailboxId(mbox.getId());
+                // cache the accountID-to-mailboxID and mailboxID-to-Mailbox relationships
+                cacheAccount(data.accountId, data.id);
+                cacheMailbox(mbox);
+                redoRecorder.setMailboxId(mbox.getId());
 
-            success = true;
+                try {
+                    if (mbox != null) {
```
Comment (spoon16) [inline store/src/java/com/zimbra/cs/mailbox/MailboxManager.java]: If `mbox == null` you are going to throw an NPE in the `try-with-resources` head. I think the logic here is wrong. Why was this moved from the finally block below? The control flow was changed. We do not want to do that.

File: store/src/java/com/zimbra/cs/service/mail/AnnounceOrganizerChange.java [hunk 38]
```diff
@@ -53,12 +54,11 @@ public Element handle(Element request, Map<String, Object> context) throws Servi
 
         MailSendQueue sendQueue = new MailSendQueue();
         Element response = getResponseElement(zsc);
-        mbox.lock.lock();
-        try {
+        try (final MailboxLock l = mbox.lock(true)) {
```
Comment (spoon16) [inline store/src/java/com/zimbra/cs/service/mail/AnnounceOrganizerChange.java]: Control flow was changed here. `l` needs to be closed before `sendQueue.send()` is called. You can manage the lifecycle of `l` yourself of your can do something like this.

```java
try {
    try (final MailboxLock l = mbox.lock(true)) {
        l.lock();
        CalendarItem calItem = mbox.getCalendarItemById(octxt, iid,getId());
        sendOrganizerChangeMessage(zsc, octxt, calItem, acct, mbox, sendQueue);
    }
} finally {
    sendQueue.send();
}
```

File: store/src/java/com/zimbra/cs/service/mail/ModifyCalendarItem.java [hunk 39]
```diff
@@ -108,8 +109,9 @@ public Element handle(Element request, Map<String, Object> context) throws Servi
         MailSendQueue sendQueue = new MailSendQueue();
         Element response = getResponseElement(zsc);
         int compNum = (int) request.getAttributeLong(MailConstants.A_CAL_COMP, 0);
-        mbox.lock.lock();
+        final MailboxLock l = mbox.lock(true);
```
Comment (spoon16) [inline store/src/java/com/zimbra/cs/service/mail/ModifyCalendarItem.java]: This is correct. It does not change control flow. An alternative to this is what I put above https://github.com/Zimbra/zm-mailbox/pull/444/files#r154731075

File: store/src/java/com/zimbra/cs/session/Session.java [hunk 40]
```diff
@@ -166,13 +167,13 @@ public Session register() throws ServiceException {
...
      * @see #isRegisteredInCache() */
-    public Session unregister() {
+    public Session unregister(MailboxLock l) {
         // locking order is always Mailbox then Session
         MailboxStore mboxStore = mailbox;
         if (null != mboxStore) {
             if (mboxStore instanceof Mailbox) {
                 Mailbox mbox = (Mailbox)mboxStore;
-                assert(mbox.lock.isWriteLockedByCurrentThread() || !Thread.holdsLock(this));
+                assert(l.isWriteLockedByCurrentThread() || !Thread.holdsLock(this));
```
Comment (spoon16) [inline store/src/java/com/zimbra/cs/session/Session.java]: _TODO_ need to figure out what to do with these calls to `Thread.holdsLock(this)`

File: store/src/java/com/zimbra/cs/mailbox/Tag.java [hunk 41]
```diff
@@ -57,7 +57,9 @@
         }
 
         NormalizedTags(Mailbox mbox, String[] tagsFromClient, boolean create, boolean imapVisible) throws ServiceException {
-            assert mbox.isTransactionActive() : "cannot instantiate NormalizedTags outside of a transaction";
```
Comment (spoon16) [inline store/src/java/com/zimbra/cs/mailbox/Tag.java]: Why did you comment this out? `isTransactionActive` seems like a valid reference.

https://github.com/Zimbra/zm-mailbox/blob/cd82ccc06afb04603e8fb9a45b5983a9b4773313/store/src/java/com/zimbra/cs/mailbox/Mailbox.java#L9981

Comment (Raffaell0) [reply store/src/java/com/zimbra/cs/mailbox/Tag.java]: isTransactionActive() method is not accesible like a direct method for mailbox reference, you need to pass for a MailboxTransaction reference, in anycase, how can you know isTransactionActive is called for what reference?

Comment (spoon16) [reply store/src/java/com/zimbra/cs/mailbox/Tag.java]: I just committed a small change that moves `isTransactionActive` back up to the `Mailbox` level. That should keep this code working.

Good catch!

File: store/src/java-test/com/zimbra/cs/mailbox/DumpsterTest.java [hunk 42]
```diff
@@ -80,8 +82,8 @@ public void flagDumpsterItem() throws Exception {
         doc = mbox.getItemById(null, doc.mId, MailItem.Type.DOCUMENT, true);
         boolean success = false;
         boolean immutableException = false;
-        try {
-            mbox.beginTransaction("alterTag", null);
+        try (final MailboxLock l = mbox.lock(true);
+        		final Mailbox.MailboxTransaction t = mbox.new MailboxTransaction("alterTag", null, l)) {
```
Comment (spoon16) [inline store/src/java-test/com/zimbra/cs/mailbox/DumpsterTest.java]: indentation is wrong

File: store/src/java-test/com/zimbra/cs/mailbox/DumpsterTest.java [hunk 43]
```diff
@@ -80,8 +82,8 @@ public void flagDumpsterItem() throws Exception {
...
             doc.alterTag(Flag.FlagInfo.FLAGGED.toFlag(mbox), true);
             success = true;
```
Comment (spoon16) [inline store/src/java-test/com/zimbra/cs/mailbox/DumpsterTest.java]: you need to call `t.commit()`

File: store/src/java-test/com/zimbra/cs/mailbox/DumpsterTest.java [hunk 44]
```diff
@@ -80,8 +82,8 @@ public void flagDumpsterItem() throws Exception {
         doc = mbox.getItemById(null, doc.mId, MailItem.Type.DOCUMENT, true);
         boolean success = false;
```
Comment (spoon16) [inline store/src/java-test/com/zimbra/cs/mailbox/DumpsterTest.java]: Variable is not needed `boolean success`. Call `t.commit()` instead

File: store/src/java-test/com/zimbra/cs/mailbox/DumpsterTest.java [hunk 45]
```diff
@@ -115,9 +115,7 @@ public void moveDumpsterDoc() throws Exception {
             if (!immutableException) {
                 throw e;
             }
-        } finally {
-            mbox.endTransaction(success);
```
Comment (spoon16) [inline store/src/java-test/com/zimbra/cs/mailbox/DumpsterTest.java]: You removed a call to `endTransaction`? But there is no corresponding transaction commit/close in the new code? Please explain why this is the correct change to make here.

File: store/src/java-test/com/zimbra/cs/mailbox/DumpsterTest.java [hunk 46]
```diff
@@ -144,9 +142,7 @@ public void moveDumpsterEmail() throws Exception {
...
```
Comment (spoon16) [inline store/src/java-test/com/zimbra/cs/mailbox/DumpsterTest.java]: You removed a call to `endTransaction`? But there is no corresponding transaction commit/close in the new code? Please explain why this is the correct change to make here.

File: store/src/java-test/com/zimbra/cs/mailbox/MailboxLockTest.java [hunk 47]
```diff
@@ -23,7 +23,7 @@
 
 import com.zimbra.cs.account.MockProvisioning;
 import com.zimbra.cs.account.Provisioning;
-import com.zimbra.common.mailbox.MailboxLock.LockFailedException;
```
Comment (spoon16) [inline store/src/java-test/com/zimbra/cs/mailbox/MailboxLockTest.java]: Why was this removed? Doesn't this exception type still exist?

File: store/src/java-test/com/zimbra/cs/mailbox/ThreaderTest.java [hunk 48]
```diff
@@ -93,14 +94,12 @@ static MimeMessage getSecondMessage() throws Exception {
     }
 
     private void threadMessage(String msg, MailThreadingAlgorithm mode, ParsedMessage pm, Mailbox mbox, List<Integer> expectedMatches) throws Exception {
-        mbox.beginTransaction("ThreaderTest", null);
-        try {
+    	try (final MailboxLock l = mbox.lock(true);
+                final Mailbox.MailboxTransaction t = mbox.new MailboxTransaction("ThreaderTest", null, l)) {
```
Comment (spoon16) [inline store/src/java-test/com/zimbra/cs/mailbox/ThreaderTest.java]: indentation/whitespace is wrong

File: store/src/java-test/com/zimbra/cs/mailbox/DumpsterTest.java (hunk 45 above)
Comment (spoon16) [reply store/src/java-test/com/zimbra/cs/mailbox/DumpsterTest.java]: Added an assertion to test whether or not there is an active transaction at this point, a quick code review makes me think that there should not be.

File: store/src/java-test/com/zimbra/cs/mailbox/DumpsterTest.java (hunk 46 above)
Comment (spoon16) [reply store/src/java-test/com/zimbra/cs/mailbox/DumpsterTest.java]: Added assertion to check to see if there is a transaction that needs to be cleaned up here.

File: store/src/java-test/com/zimbra/cs/mailbox/DumpsterTest.java (hunk 45 above)
Comment (Raffaell0) [reply store/src/java-test/com/zimbra/cs/mailbox/DumpsterTest.java]: I removed mbox.endTransaction(success) because new implementarion calls close implicitly, but you're right, I forgot to call commit method.
